src/v/cluster/client_quota_serde.h

```diff
@@ -164,32 +166,117 @@ struct entity_value
+    friend std::ostream& operator<<(std::ostream&, const entity_value_diff&);
+
+    auto serde_fields() { return std::tie(entries); }
+
+    absl::flat_hash_set<entry> entries;
+};
+
 struct alter_delta_cmd_data
   : serde::envelope<
       alter_delta_cmd_data,
-      serde::version<0>,
-      serde::compat_version<0>> {
-    struct upsert_op
-      : serde::
-          envelope<upsert_op, serde::version<0>, serde::compat_version<0>> {
+      serde::version<1>,
+      serde::compat_version<1>> {
+    struct op
+      : serde::envelope<op, serde::version<0>, serde::compat_version<0>> {
         client_quota::entity_key key;
-        client_quota::entity_value value;
-        auto serde_fields() { return std::tie(key, value); }
+        client_quota::entity_value_diff diff;
+        auto serde_fields() { return std::tie(key, diff); }
     };
 
-    struct remove_op
-      : serde::
-          envelope<remove_op, serde::version<0>, serde::compat_version<0>> {
-        client_quota::entity_key key;
-        auto serde_fields() { return std::tie(key); }
-    };
-
-    std::vector<upsert_op> upsert;
-    std::vector<remove_op> remove;
+    std::vector<op> ops;
 
     friend bool
     operator==(const alter_delta_cmd_data&, const alter_delta_cmd_data&)
       = default;
 };
 
 } // namespace cluster::client_quota
+
+constexpr std::string_view
+to_string_view(cluster::client_quota::entity_value_diff::key e) {
+    /// Note: the string values of these enums need to match the values used in
+    /// the kafka client quota handlers
+    switch (e) {
+    case cluster::client_quota::entity_value_diff::key::producer_byte_rate:
+        return "producer_byte_rate";
+    case cluster::client_quota::entity_value_diff::key::consumer_byte_rate:
+        return "consumer_byte_rate";
+    case cluster::client_quota::entity_value_diff::key::
+      controller_mutation_rate:
+        return "controller_mutation_rate";
+    }
+}
+
+template<typename E>
+std::enable_if_t<std::is_enum_v<E>, std::optional<E>>
+  from_string_view(std::string_view);
+
+template<>
+constexpr std::optional<cluster::client_quota::entity_value_diff::key>
+from_string_view<cluster::client_quota::entity_value_diff::key>(
+  std::string_view v) {
+    return string_switch<
+             std::optional<cluster::client_quota::entity_value_diff::key>>(v)
+      .match(
+        to_string_view(
+          cluster::client_quota::entity_value_diff::key::producer_byte_rate),
+        cluster::client_quota::entity_value_diff::key::producer_byte_rate)
+      .match(
+        to_string_view(
+          cluster::client_quota::entity_value_diff::key::consumer_byte_rate),
+        cluster::client_quota::entity_value_diff::key::consumer_byte_rate)
+      .match(
+        to_string_view(cluster::client_quota::entity_value_diff::key::
+                         controller_mutation_rate),
+        cluster::client_quota::entity_value_diff::key::controller_mutation_rate)
+      .default_match(std::nullopt);
+}
```

Comment:
Can these be moved to the namespace, please.

Reply:
Could have dropped the `cluster::client_quota` from the types, too.

Reply:
I'll add that to the follow up PR